### PR TITLE
RST-717 Correctly handling state restoration.

### PR DESCRIPTION
**Pull request by ayrton04**
Addresses https://locusrobotics.atlassian.net/browse/RST-717 by making sure that we don't return from the `revertTo` method prematurely when the state history isn't long enough.

### CODE REVIEW

File: src/ros_filter.cpp [hunk 1]
```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
 
     // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
     // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
-    // the queue.
+    // the queue. If the history is insufficiently short, we just take the oldest state we have.
+    FilterStatePtr lastHistoryState;
     while (!filterStateHistory_.empty() && filterStateHistory_.back()->lastMeasurementTime_ > time)
```
Comment (svwilliams) [inline src/ros_filter.cpp]: This is all working, so I'm not suggesting you change anything. But would it have been simpler just to change `while` loop condition to `(filterStateHistory_.size() > 1) && (filterStateHistory_.back()->lastMeasurementTime_ > time)` to prevent clearing the entire `filterStateHistory_` container?

File: src/ros_filter.cpp [hunk 2]
```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
...
+      lastHistoryState = filterStateHistory_.back();
       filterStateHistory_.pop_back();
     }
 
+    // If the state history is not empty at this point, it means that our history was large enough, and we
+    // should revert to the state at the back of the history deque.
+    bool retVal = false;
+    if (!filterStateHistory_.empty())
+    {
+      retVal = true;
```
Comment (svwilliams) [inline src/ros_filter.cpp:2989]: You are reverting to a previous time regardless of whether you are reverting to the desired previous time or not, but you are only returning "true" if you reverted to the desired time. I just want to verify that is the desired meaning of the return code.

File: src/ros_filter.cpp [hunk 3]
```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
...
+      lastHistoryState = filterStateHistory_.back();
```
Comment (svwilliams) [inline src/ros_filter.cpp:2990]: In both the original code and in this PR, in the case where the you had sufficient history, you are setting the `lastHistoryState` to a value that is still in the `filterStateHistory_` container. From this, I am assuming that (a) you will rebuild the `filterStateHistory_` entries from the `measurementQueue_` in some other function, and that (b) the current filter state will not be re-added to the `filterStateHistory_` since it still exists in that container.

However, in the case where there is not sufficient history, the value of `lastHistoryState` does not exist in the `filterStateHistory_`. Does this mean you end up dropping that first state from the `filterStateHistory_` container during the history rebuild?

**Review by svwilliams — APPROVED** [review]
Approved...with some questions.

File: src/ros_filter.cpp (hunk 2 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2989]: Correct. The semantics of the return type are a little weird, I'll grant you.

File: src/ros_filter.cpp (hunk 3 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2990]: To back way up, before this state reversion code existed, if we got a measurement that was old, we just issued a warning, skipped the prediction step (since the time delta would have been negative), and then corrected.

The `filterStateHistory` only gets updated when we fuse a measurement. We save the state at the time the measurement was fused.

In the original code for this PR, we only bothered reverting the state if there was sufficient history, so we'd back up both the measurement and state histories (which should be parallel) until we reached a time that was <= the one we requested. We would *not* re-enqueue the measurement that was <= the time in question, though, so the `filterStateHistory` would not get any duplicate entries.

When we have insufficient history, it means that we would like to have reverted to some state, but we can't, so we just revert as far back as we can go, and then pass those measurements through, starting with the measurement that made us have to revert to begin with. When we get to that measurement, which is older than any state or measurement we have, we end up with a situation that is pretty much identical to what I described in the first paragraph.

TL; DR, in the first case (sufficient history) we are reverting to the right point in history and then replacing the history values *after* that time as the measurements are processed. In the second case, we are reverting to the oldest point in history we can (which clears the history), and then fusing all the measurements again, which will rebuild the history from the time at which we reverted the state.

I wish I had a whiteboard.

File: src/ros_filter.cpp (hunk 1 above)
Comment (ayrton04) [reply src/ros_filter.cpp]: So the premise is that, if the measurement we get is older than the oldest state in our history, we want to replace the oldest state in our history. If the history is too short, we will clear it, and then when we try to fuse the old measurement that prompted the reversion, we will do an immediate correction at the time in question, and that will become our new state for that time, and it will get stored in the history. If I don't clear the history when its length is insufficient (i.e., if we keep the one at the back), we'll have two states with the same time stamp after we process the "old" measurement.

Note that this really only will affect us within the first 0.4 seconds of execution (current history length), after which time we will always have sufficient history.

File: src/ros_filter.cpp [hunk 4]
```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
 
     // Walk back through the queue until we reach a filter state whose time stamp is less than or equal to the
     // requested time. Since every saved state after that time will be overwritten/corrected, we can pop from
-    // the queue.
+    // the queue. If the history is insufficiently short, we just take the oldest state we have.
```
Comment (tappan-at-git) [inline src/ros_filter.cpp]: Should this be either "too short" or "insufficiently long" (far back)?

File: src/ros_filter.cpp (hunk 3 above)
Comment (svwilliams) [reply src/ros_filter.cpp:2990]: I don't want to belabor this point too long, but...

Based on your explanation, then the timestamps associated with the `measurementHistory_`/`measurementQueue_` and the `filterStateHistory_` are synchronized. I am going to assume the `filterStateHistory_` represents the filter state *after* the measurement has been applied.

So, if I receive measurements at t=1 and t=2. `measurementHistory_` contains two entries (for t=1, and t=2), and `filterStateHistory_` also contains 2 entries (t=1, and t=2). If I call `revertTo(1.5s)`, then I reset the state to the value before t=1.5, i.e. the cached state at t=1, and I queue up all measurements received after t=1.5, i.e. just the measurement at t=2. You do not queue up the measurement received at t=1.

However, if I call `revertTo(0.5s)`, then I set the state to the oldest state in `filterStateHistory_`, i.e. the state at time t=1, and a queue up all measurements after t=0.5s, i.e. the measurement at t=1 and the measurement at t=2. Assuming that all of the measurements in the queue will get "replayed", the measurement at t=1 will get applied twice, once when the state was generated and a second time when the measurement was replayed.

Granted, I have not looked at the measurement replay code. So there may be logic that handles such a situation.

File: src/ros_filter.cpp [hunk 5]
```diff
@@ -2971,12 +2973,32 @@ namespace RobotLocalization
...
+    // should revert to the state at the back of the history deque.
+    bool retVal = false;
+    if (!filterStateHistory_.empty())
+    {
+      retVal = true;
+      lastHistoryState = filterStateHistory_.back();
+    }
+    else
+    {
+      RF_DEBUG("Insufficient history to revert to time " << time << "\n");
```
Comment (tappan-at-git) [inline src/ros_filter.cpp:2994]: Would it be more better to check whether `filterStateHistory.front()->lastMeasurementTime_ < time` to skip the individual pop & test? If this is happening a lot, it might be meaningfully more efficient. Otherwise I like your current design.

**Review by tappan-at-git — APPROVED** [review]

File: src/ros_filter.cpp (hunk 3 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2990]: So the short answer is that yes, the `t1` measurement will get replayed over the state for which it itself was responsible for generating.

However, if we're calling `revertTo(0.5)`, it means we have a measurement `m_old` that arrived with a time stamp of `0.5`. So when we get back to processing measurements, the measurement queue will have three measurements: `m_old`, and the "replayed" measurements from `t1` and `t2`.  The EKF will do an immediate correction on the oldest state using `m_old`, *then* replay the measurement at `t1`, then the measurement at `t2`. So I guess we'd end up with a state that was basically prior state, followed by measurement `t1`, then `m_old`, then measurement `t1` again, then measurement `t2`.

Is there a clean solution to this? Even if I were to revert such that we *didn't* replay the `t1` measurement (and didn't pop the oldest state), we'd still have a state that ended up being a result of fusing `t1`, then `m_old`, then `t2`. `m_old` would still be out of order. I could go either way, really. I'm also not really that concerned, since (a) this would only happen, if at all, once at the beginning of execution (depending on the history length), and (b) it's a degenerate case that I feel users should be responsible for handling.

What do you think?

File: src/ros_filter.cpp (hunk 5 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2994]: Most of the logic for this history management (including that) came from a contributor PR, actually.

I can't check the `front`, necessarily, because then I'd have to pop to be able to look at the next value, and I don't want to do that. This is a `deque`, though, which supports random access, so we can walk to the point in question and then do an `erase` on the rest. However, I wouldn't want to start at the `0` index/front because we're more likely to be forced to search through more results to get to the time in question than if we start from the back of the deque and work towards the front.

File: src/ros_filter.cpp (hunk 4 above)
Comment (ayrton04) [reply src/ros_filter.cpp]: Yep.

File: src/ros_filter.cpp (hunk 5 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2994]: However, it looks like `pop_back` is a constant-time operation, so given that this is working, I'm inclined to leave it alone, unless you feel strongly about it.

File: src/ros_filter.cpp (hunk 3 above)
Comment (svwilliams) [reply src/ros_filter.cpp:2990]: > What do you think?

It's weird. Given that reverting to a time before the beginning of the window should not happen, the details probably don't matter too much.

In my example, you are applying a measurement that occurred at time t=0.5s to the state estimate from time t=1.0s. I don't think anything you do could make that right. I do believe you are "double counting" that one measurement. In general measurements that happen at the same time can be applied to the state estimate in any order. So, in the limit, if we revert to t=0.99999999s, then we would have already applied the measurement at t=1.0. If we then apply just the measurement at t=0.9999999999s we should get the same answer as if we applied the measurements in the opposite, correct order. Apply the measurement at t=1.0s, then the measurement at t=0.999999999s, then the measurement at t=1.0s should give a different answer. For whatever that is worth.

Feel free to merge whenever you are tired of this conversation. :)

Comment (svwilliams) [reply src/ros_filter.cpp:2990]: Also, what was the motivation for including these measurements that are outside of the history window? Instead of the old approach that just ignored them? I mean, I could give you a measurement that was an hour old. If you include that, you will completely mangle the state.

File: src/ros_filter.cpp (hunk 5 above)
Comment (tappan-at-git) [reply src/ros_filter.cpp:2994]: I don't feel strongly about it, no. And to be clear, this suggestion is only as an alternative to running through every item in the queue to decide if we have a long enough history -- I definitely agree that working backwards from the most recent measurement is more sensible for finding the last relevant measurement.

File: src/ros_filter.cpp (hunk 3 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2990]: What I find is that if I ignore them, then I get ROS Answers questions like "OMG why is my measurement for sensor XYZ being ignored?" It's usually because they have a low-latency sensor mixed with a high-latency one. I could define a threshold, but then it feels kinda' arbitrary (not unlike the way I am handling it now).

Yes, I fully agree that I am double-counting that measurement, but I'm of the mind that this is the corneriest of corner cases for any reasonable data set, and one that I'm comfortable with, unless you feel strongly about it.

File: src/ros_filter.cpp (hunk 5 above)
Comment (ayrton04) [reply src/ros_filter.cpp:2994]: Oooh, I see, you mean just doing an immediate check on the length of the queue, gotcha.

Comment (tappan-at-git) [reply src/ros_filter.cpp:2994]: Yeah. You have the option of just looking at how much time the queue spans directly. I don't know how much of a benefit that is, but I guess the more inputs you have (and the higher frequency they are) the more pops you may have to do unnecessarily.

Comment (ayrton04) [reply src/ros_filter.cpp:2994]: But that would only be true if the queue is too short, right? Once it fills up even a little bit, we'll almost never get a measurement that exceeds its bounds. For example, we have it set to 0.4 seconds, and we violate it maybe once in the entire lifetime of the node (right at the beginning of execution).

Also, a `clear` on the deque is [O(n)](http://en.cppreference.com/w/cpp/container/deque/clear), as is the act of popping every value in it. I'm sure popping is more like 2n, but I'm personally not fussed, so long as they're both linear.

Comment (tappan-at-git) [reply src/ros_filter.cpp:2994]: Yeah, definitely sounds like it wouldn't be worth it.
